bitswap.go

```diff
@@ -322,23 +317,26 @@ func (bs *Bitswap) receiveBlocksFrom(ctx context.Context, from peer.ID, blks []b
 	default:
 	}
 
-	wanted := blks
-
-	// If blocks came from the network
-	if from != "" {
-		var notWanted []blocks.Block
-		wanted, notWanted = bs.sim.SplitWantedUnwanted(blks)
-		for _, b := range notWanted {
-			log.Debugf("[recv] block not in wantlist; cid=%s, peer=%s", b.Cid(), from)
-		}
+	wantedKs, err := bs.wrm.PublishToSessions(&bswrm.IncomingMessage{
+		From:      from,
+		Blks:      blks,
+		Haves:     haves,
+		DontHaves: dontHaves,
+	})
+	if err != nil {
+		return err
 	}
 
-	// Put wanted blocks into blockstore
-	if len(wanted) > 0 {
-		err := bs.blockstore.PutMany(wanted)
```

Comment:
blockstore operations now happen inside of the WantRequestManager so it can synchronize access